src/tcp.rs

```diff
@@ -79,16 +94,36 @@ pub async fn connect(
         };
 
         configure_socket(socket2::SockRef::from(&socket), &so_mark)?;
-        match timeout(connect_timeout, socket.connect(addr)).await {
+
+        // Spawn the connection attempt in the join set.
+        // We include a delay of ix * 250 milliseconds, as per RFC8305.
+        // See https://datatracker.ietf.org/doc/html/rfc8305#section-5
+        join_set.spawn(sleep(Duration::from_millis(250 * (ix as u64))).then(move |_| {
+            timeout(connect_timeout, socket.connect(addr).map_err(move |err| (addr, err)))
+                .map_err(move |err| (addr, err))
+        }));
+    }
+
+    // Wait for the next future that finishes in the join set, until we got one
+    // that resulted in a successful connection.
+    while let Some(res) = join_set.join_next().await {
+        match res? {
             Ok(Ok(stream)) => {
+                // We've got a successful connection, so we can abort all other
+                // on-going attempts.
+                join_set.shutdown().await;
```

Comment:
You don't need to shutdown and await, when the joinset is going to be dropped at the function exit, everything is going to aborted when the object is drop.

if you want to be explicit only abort_all() is needed, not awaiting for the cancellation to terminate

Reply:
Ok, I had to add an additional check to the while loop condition so that we break out of the loop as soon as `cnx` is no longer `None`.